clients/java/src/main/java/io/camunda/zeebe/client/impl/command/CreateDocumentCommandImpl.java

```diff
@@ -0,0 +1,159 @@
+import io.camunda.zeebe.client.impl.response.DocumentReferenceResponseImpl;
+import io.camunda.zeebe.client.protocol.rest.DocumentMetadata;
+import io.camunda.zeebe.client.protocol.rest.DocumentReference;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
+import org.apache.hc.client5.http.entity.mime.StringBody;
+import org.apache.hc.core5.http.ContentType;
+
+@ExperimentalApi("https://github.com/camunda/issues/issues/841")
+public class CreateDocumentCommandImpl
+    implements CreateDocumentCommandStep1, CreateDocumentCommandStep2 {
+
+  private String documentId;
+  private String storeId;
+  private InputStream content;
+  private final DocumentMetadata metadata;
+  private final JsonMapper jsonMapper;
+  private final HttpClient httpClient;
+  private final RequestConfig.Builder httpRequestConfig;
+
+  public CreateDocumentCommandImpl(final JsonMapper jsonMapper, final HttpClient httpClient) {
+    metadata = new DocumentMetadata();
+    this.jsonMapper = jsonMapper;
+    this.httpClient = httpClient;
+    httpRequestConfig = httpClient.newRequestConfig();
+  }
+
+  @Override
+  public FinalCommandStep<DocumentReferenceResponse> requestTimeout(final Duration requestTimeout) {
+    httpRequestConfig.setResponseTimeout(
+        requestTimeout.toMillis(), java.util.concurrent.TimeUnit.MILLISECONDS);
+    return this;
+  }
+
+  @Override
+  public ZeebeFuture<DocumentReferenceResponse> send() {
+    final MultipartEntityBuilder entityBuilder =
+        MultipartEntityBuilder.create().setContentType(ContentType.MULTIPART_FORM_DATA);
+
+    final String name = Optional.ofNullable(metadata.getFileName()).orElse("");
+    entityBuilder.addBinaryBody("file", content, ContentType.DEFAULT_BINARY, name);
+
+    final String metadataString = jsonMapper.toJson(metadata);
+    entityBuilder.addPart("metadata", new StringBody(metadataString, ContentType.APPLICATION_JSON));
+
+    final HttpZeebeFuture<DocumentReferenceResponse> result = new HttpZeebeFuture<>();
+
+    final Map<String, String> queryParams = new HashMap<>();
+    if (documentId != null) {
+      queryParams.put("documentId", documentId);
+    }
+    if (storeId != null) {
+      queryParams.put("storeId", storeId);
+    }
+    httpClient.postMultipart(
+        "/documents",
+        queryParams,
+        entityBuilder,
+        httpRequestConfig.build(),
+        DocumentReference.class,
+        DocumentReferenceResponseImpl::new,
+        result);
+    return result;
+  }
+
+  @Override
+  public CreateDocumentCommandStep2 content(final InputStream content) {
+    this.content = content;
+    return this;
+  }
+
+  @Override
+  public CreateDocumentCommandStep2 content(final byte[] content) {
```

Comment:
I think the role of the file name in the document creation process is a bit different. For the document object, a file name is just an optional part of the metadata - we are anyway generating an ID for each file and not using the file name as an identifier. We have to add it to the multipart request too, though, because Spring MVC doesn't support working with request parts as files if they have no filename. But we want to allow the users to create documents without a file name.

Reply:
Thanks for sharing, it make sense. Do we need to close the `InputStream` then ? We can add a try/finally state in the send command and close the stream there

Reply:
Good point!